### ISSUE TITLE
Rpc: possible performance improvement for filters?

### ISSUE THREAD

**Issue by fanatid** [solana-labs/solana#20177]
#### Problem

https://github.com/solana-labs/solana/blob/ca83167cfcc4f5cdb0837bf04d99041e7377c516/client/src/rpc_filter.rs#L49-L53

`MemcmpEncodedBytes` enum looks not healthy, why `Binary` variant has `String` as inner? And more, this `String` have bytes encoded in base58 :anguished: 

I'm not sure about the performance impact but decode base58 string each time for filter account on requests like `get_program_accounts` looks not good.

#### Proposed Solution

1. Rename `Binary` to `Base58` and add `Binary(Vec<u8>)`.
2. Left `Binary(String)` and add new variant with `Vec<u8>` (name?).

**Comment by mvines** [comment]
I agree it would be nice to have a new `Base64(String)` variant of this enum to avoid the base58 entirely.  We can't change the existing `Binary()` name as that'd would break backwards compatibility.

But it's worse!  The `Binary(String)` shouldn't be terrible if the base58 decode only occurs once per RPC call, but this appears to not be the case:

`bytes_match()` is called for each program account:
https://github.com/solana-labs/solana/blob/122206dbb19075358b5bb785cd1056f7e0f4d899/rpc/src/rpc.rs#L1799
and it internally performs a base58 decode into a local variable
https://github.com/solana-labs/solana/blob/122206dbb19075358b5bb785cd1056f7e0f4d899/client/src/rpc_filter.rs#L69
...every time. ☠️ 

It should be a big win to memoize that base58 decode so it only occurs once per filter per RPC request

**Comment by fanatid** [comment]
If we add `BinaryUnderOtherName(Vec<u8>)` we would able to convert `Base64(String)` / `Base58(String)` / etc to decoded variant before applying filter. Just not sure under which name to add `Vec<u8>` :laughing: .

**Comment by mvines** [comment]
A `Vec<u8>` over JSON RPC would be encoded as an array of numbers, which isn't super efficient bandwidth-wise.  That's why we favour encodings that end up as a string over the wire generally

**Comment by fanatid** [comment]
Thanks for clearing this up. I'll try to solve this issue.
